Resource generator --use-model #297 fixes #328

as27 opened the pull request:
Generates a basic implementation for a resource, which supports crud functionality for a existing model.

markbates commented:
I ran `buffalo g resource user name email bio:text`

it's output:

```
Buffalo version 0.8.0

--> actions/user.go
--> actions/user_test.go
--> buffalo db g model user name email bio:text
v3.13.3

--> models/user.go
--> models/user_test.go
--> goimports -w .
> migrations/20170327183424_create_users.up.fizz
> migrations/20170327183424_create_users.down.fizz
--> goimports -w .
```

The `UserResource`, however, didn't have any of the model stuff in it. It should.

```go
package actions

import "github.com/gobuffalo/buffalo"

type UserResource struct {
	buffalo.Resource
}

// List default implementation.
func (v UserResource) List(c buffalo.Context) error {
	return c.Render(200, r.String("User#List"))
}

// Show default implementation.
func (v UserResource) Show(c buffalo.Context) error {
	return c.Render(200, r.String("User#Show"))
}

// New default implementation.
func (v UserResource) New(c buffalo.Context) error {
	return c.Render(200, r.String("User#New"))
}

// Create default implementation.
func (v UserResource) Create(c buffalo.Context) error {
	return c.Render(200, r.String("User#Create"))
}

// Edit default implementation.
func (v UserResource) Edit(c buffalo.Context) error {
	return c.Render(200, r.String("User#Edit"))
}

// Update default implementation.
func (v UserResource) Update(c buffalo.Context) error {
	return c.Render(200, r.String("User#Update"))
}

// Destroy default implementation.
func (v UserResource) Destroy(c buffalo.Context) error {
	return c.Render(200, r.String("User#Destroy"))
}
```

markbates commented on generators/resource/templates/actions/resource-use_model.go.tmpl:
```diff
@@ -0,0 +1,166 @@
+func (v {{.camel}}Resource) List(c buffalo.Context) error {
+	// Get the DB connection from the context
+	tx := c.Value("tx").(*pop.Connection)
+	{{.downFirstCap}} := &models.{{.camel}}{}
```
This should be plural.

It's generating `user := &models.User{}`, but it should be `users := &models.Users{}`

as27 replied:
I thought that resources are always generated with plural args: 'buffalo g resource users name email bio:text' then the logic works. But I am going to change that.

as27 replied:
The model stuff is just implemented with the `--use-model` flag. Because otherwise the tests will fail.

markbates replied:
Then the tests need to be fixed. The `--use-model` flag should be used if you have an existing model, but the default, without any flags, should generate everything.

It's ok in the Dockerfile to delete stuff that was previously generated in the tests, or to create a new app, etc...

markbates commented:
This should generate all of the `*.html` files as well. Each file can contain something basic like:

```html
<h1>Users#List</h1>
Find me at templates/users/list.html
```

as27 commented:
* Logic of the flags changed
* added a basic structure for html templates. Later just the files can hold more logic
* changed the generator also to deal with a singular definition.

markbates commented on generators/resource/templates/actions/resource-use_model.go.tmpl:
```diff
@@ -0,0 +1,172 @@
+	// If there are no errors set a success message
+	c.Flash().Add("success", "{{.model}} was created successfully")
+	// and redirect to the {{.underPlural}} index page
+	return c.Redirect(301, "/{{.underPlural}}")
```
Have this redirect to the `Show` action. `/users/1234`

markbates replied:
I think this should be a `302`.

markbates commented on generators/resource/templates/actions/resource-use_model.go.tmpl:
```diff
@@ -0,0 +1,172 @@
+	// If there are no errors set a success message
+	c.Flash().Add("success", "{{.model}} was edited successfully")
+	// and redirect to the {{.underPlural}} index page
+	return c.Redirect(301, "/{{.underPlural}}")
```
Have this redirect to the `Show` action. `/users/1234`

markbates replied:
I think this should be a `302`.

markbates commented on generators/resource/templates/actions/resource-use_model.go.tmpl:
```diff
@@ -0,0 +1,172 @@
+	if err != nil {
+		return err
+	}
+	// Redirect to the {{.underPlural}} index page
```
Add a flash message here to say that it was deleted.

markbates commented on generators/resource/templates/actions/resource-use_model.go.tmpl:
```diff
@@ -0,0 +1,172 @@
+		return err
+	}
+	// Redirect to the {{.underPlural}} index page
+	return c.Redirect(301, "/{{.underPlural}}")
```
I think this should be a `302`.

markbates commented on buffalo/cmd/generate/resource.go:
```diff
@@ -31,14 +31,23 @@ $ buffalo g resource users --skip-model
 var SkipResourceModel = false
 
+// UseResourceModel allows to generate a resource with a working model.
+var UseResourceModel = false
```
this should be a `string` containing the name of the model to use.

```
buffalo g resource users --use-model=user
```

markbates commented:
Some issues with pluralization:

```
buffalo g resource user name email
```

```
--> actions/user.go
--> actions/user_test.go
--> templates/users/_form.html
--> templates/users/edit.html
--> templates/users/index.html
--> templates/users/new.html
--> templates/users/show.html
--> buffalo db g model user name email
v3.13.5

--> models/user.go
--> models/user_test.go
--> goimports -w .
> migrations/20170329204449_create_users.up.fizz
> migrations/20170329204449_create_users.down.fizz
--> goimports -w .
```

In `actions/app.go`:

```go
var userResource buffalo.Resource
userResource = UserResource{&buffalo.BaseResource{}}
app.Resource("/user", userResource)
```

But the resource is named `UsersResource`.

markbates commented:
It looks like the happy path is working great!

```
$ buffalo g resource users field1 field2 ...
$ buffalo g resource --skip-model users
```

These still have issues:

```
$ buffalo g resource user field1 field2 ...
$ buffalo g resource --use-model users
```

as27 commented:
* Changed template according to the review
* Minor changes to support different values for resource and model: `buffalo g r users --use-model people`
* Supports longer camel case names for example: UserComments

as27 commented:
Basic set of html templates are now generated.
It supports list, show, edit, new and delete out of the box.

markbates commented:
https://github.com/gobuffalo/buffalo/pull/352

I fixed a few bugs and cleaned up some of the HTML.
